### PR TITLE
Split smoothing length iteration from density calculation.

### PR DESCRIPTION
**Pull request by rainwoodman**
- stablizes the presure-entropy iteration at init
 - equivalent of doing one extra iteration for density
 - but each iteration in sml we don't compute additional junk for
   div etc.

 - might be useful for gizmo like schemes.

### CODE REVIEW

**Comment by sbird** [comment]
I wonder how much work it would be to add a unit test that the smoothing length is correctly calculated? For, say, a uniformly distributed set of particles, as in the forcetree test?

File: density.c [hunk 1]
```diff
@@ -126,66 +123,7 @@ void density(void)
...
-        const int p_i = ActiveParticle[i];
-        P[p_i].DensityIterationDone = 0;
-        DENSITY_GET_PRIV(tw)->Left[p_i] = 0;
-        DENSITY_GET_PRIV(tw)->Right[p_i] = 0;
-
-    }
-
-    /* allocate buffers to arrange communication */
-
-    walltime_measure("/SPH/Density/Init");
```
Comment (sbird) [inline density.c]: Can you change the names of the walltime indicators in sml to read something else? eg, "SPH/SML/Init"

File: density.c [hunk 2]
```diff
@@ -408,121 +340,4 @@ density_postprocess(int i, TreeWalk * tw)
...
-
-void density_check_neighbours (int i, TreeWalk * tw) {
-    /* now check whether we had enough neighbours */
-
-    double desnumngb = All.DesNumNgb;
-
-#ifdef BLACK_HOLES
-    if(P[i].Type == 5)
-        desnumngb = All.DesNumNgb * All.BlackHoleNgbFactor;
-#endif
```
[the remark this answers is not in the document]
Comment (rainwoodman) [reply density.c]: No we can't . SML computes the smoothing length (feedback radius) of the BH. but BH still need an up to date density estimation for the feedback.

**Comment by rainwoodman** [comment]
Currently this is making density / sml a lot slower. on a lya run,

static tree uses 203 seconds
this uses 120(sml) + 160(density)  = 280 seconds. ~ 40% slower. 

The last commit may be able reduce this slow down. I am waiting to see.

**Comment by rainwoodman** [comment]
No it didn't help. It is probably due to the overhead of starting a new tree walk.

The main purpose of this PR is to fix initialization of pressure-entropy.

An alternative is to add a parameter to density() which does not recompute the sml. This is called during initialization.

**Pull request closed without merging**